NetCDF subdataset with multiple colons is unopenable since 3.13.0

### What is the bug?

Rasterio's CI has a failure with the latest version of GDAL. The failure involves opening a netCDF subdataset that is in a file whose name includes a colon.

Here's an excerpt from the Rasterio tests:

```python
   raise exc
E   rasterio._err.CPLE_OpenFailedError: netcdf:"/tmp/pytest-of-root/pytest-54/test_subdatasets__parsing_colo0/RGB_2026-01-01T00:00:00.nc":Band1: No such file or directory

rasterio/_err.pyx:359: CPLE_OpenFailedError

During handling of the above exception, another exception occurred:

tmp_path = PosixPath('/tmp/pytest-of-root/pytest-54/test_subdatasets__parsing_colo0')

    @pytest.mark.skipif(
        not HAVE_NETCDF or platform.system() == "Windows",
        reason="GDAL not compiled with NetCDF driver or Windows machine.",
    )
    def test_subdatasets__parsing_colons(tmp_path):
        """Ensure paths with ':' can be opened as subdatasets"""
        test_file = tmp_path / "RGB_2026-01-01T00:00:00.nc"
        shutil.copy("tests/data/RGB.nc", test_file)
        with rasterio.open(test_file) as src:
            subs = src.subdatasets
            assert len(subs) == 3
            for path in subs:
                assert path.startswith("netcdf")
>               with rasterio.open(path):
                     ^^^^^^^^^^^^^^^^^^^

tests/test_subdatasets.py:34:
```

This looks like a regression. A return to the state of GDAL when #8869 was opened.

I have a non-Python reproducer that uses one of Rasterio's test files.

### Steps to reproduce the issue

In a GDAL `ubuntu-full-3.13.0` container, run

```
root@19dcd1438b86:/app# python -c "import urllib.request; urllib.request.urlretrieve('https://github.com/rasterio/rasterio/raw/refs/heads/main/tests/data/RGB.nc', '/tmp/RGB_2026-07-24T00:00:00.nc')"
root@19dcd1438b86:/app# gdalinfo '/tmp/RGB_2026-07-24T00:00:00.nc'
Driver: netCDF/Network Common Data Format
Files: /tmp/RGB_2026-07-24T00:00:00.nc
Size is 512, 512
Metadata:
  NC_GLOBAL#Conventions=CF-1.5
  NC_GLOBAL#GDAL=GDAL 2.2.1, released 2017/06/23
  NC_GLOBAL#history=Mon Aug 28 23:28:17 2017: GDAL Create( /private/tmp/RGB.nc, ... )
Subdatasets:
  SUBDATASET_1_NAME=NETCDF:"/tmp/RGB_2026-07-24T00:00:00.nc":Band1
  SUBDATASET_1_DESC=[718x791] Band1 (8-bit integer)
  SUBDATASET_2_NAME=NETCDF:"/tmp/RGB_2026-07-24T00:00:00.nc":Band2
  SUBDATASET_2_DESC=[718x791] Band2 (8-bit integer)
  SUBDATASET_3_NAME=NETCDF:"/tmp/RGB_2026-07-24T00:00:00.nc":Band3
  SUBDATASET_3_DESC=[718x791] Band3 (8-bit integer)
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,  512.0)
Upper Right (  512.0,    0.0)
Lower Right (  512.0,  512.0)
Center      (  256.0,  256.0)
root@19dcd1438b86:/app# gdalinfo 'NETCDF:"/tmp/RGB_2026-07-24T00:00:00.nc":Band1'
ERROR 4: NETCDF:"/tmp/RGB_2026-07-24T00:00:00.nc":Band1: No such file or directory
gdalinfo failed - unable to open 'NETCDF:"/tmp/RGB_2026-07-24T00:00:00.nc":Band1'.
```

There's no such problem with a `ubuntu-full-3.12.4` image, I'm able to see all info about `Band1`.

### Versions and provenance

GDAL 3.13.0 and 3.13.1.

### Additional context

_No response_

cc @elpaso

--- a/autotest/cpp/test_cpl.cpp
+++ b/autotest/cpp/test_cpl.cpp
@@ -497,6 +497,19 @@ TEST_F(test_cpl, CSLTokenizeString2)
         auto oIteratorWrapper = cpl::IterateNameValue(papszList);
         EXPECT_TRUE(oIteratorWrapper.begin() == oIteratorWrapper.end());
     }
+
+    {
+        // Test colon separator within string tokens
+        CPLStringList aosList(CSLTokenizeString2(
+            R"(one:"two:two_and_half":three:"four:four_and_half":five)", ":",
+            CSLT_HONOURSTRINGS | CSLT_PRESERVEQUOTES));
+        EXPECT_EQ(aosList.size(), 5);
+        EXPECT_STREQ(aosList[0], "one");
+        EXPECT_STREQ(aosList[1], "\"two:two_and_half\"");
+        EXPECT_STREQ(aosList[2], "three");
+        EXPECT_STREQ(aosList[3], "\"four:four_and_half\"");
+        EXPECT_STREQ(aosList[4], "five");
+    }
 }
 
 typedef struct

--- a/autotest/gcore/hdf4_read.py
+++ b/autotest/gcore/hdf4_read.py
@@ -534,6 +534,26 @@ def test_gdal_subdataset_get_filename(filename, path_component):
 @pytest.mark.parametrize(
     "filename",
     (
+        "/vsimem/DATA:TEST.hdf",
+        "/vsimem/DATA:TEST:DOUBLE.hdf",
+        '/vsimem/DATA:"TEST".hdf',
+    ),
+)
+def test_gdal_subdataset_get_filename_colon(filename):
+    """Test regression #14978"""
+
+    gdal.FileFromMemBuffer(filename, open("data/General_RImages.hdf", "rb").read())
+    info = gdal.GetSubdatasetInfo(
+        f'HDF4_EOS:EOS_SWATH:"{filename}":/navigation_data/latitude'
+    )
+    assert info.GetPathComponent() == filename
+    assert info.GetSubdatasetComponent() == "/navigation_data/latitude"
+    gdal.Unlink(filename)
+
+
+@pytest.mark.parametrize(
+    "filename",
+    (
         'HDF4_EOS:EOS_SWATH:"AMSR_E_L2_Ocean_B01_200206182340_A.hdf":Swath1:Low_res_sst',
         r'HDF4_EOS:EOS_SWATH:"C:\AMSR_E_L2_Ocean_B01_200206182340_A.hdf":Swath1:Low_res_sst',
         "HDF4_EOS:EOS_SWATH:AMSR_E_L2_Ocean_B01_200206182340_A.hdf:Swath1:Low_res_sst",

--- a/autotest/gcore/hdf4_read.py
+++ b/autotest/gcore/hdf4_read.py
@@ -491,34 +491,9 @@ def test_hdf4_read_online_11():
     "filename,path_component",
     (
         (
-            'HDF4_EOS:EOS_SWATH:"AMSR_E_L2_Ocean_B01_200206182340_A.hdf":Swath1:Low_res_sst',
-            "AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
-        ),
-        (
-            r'HDF4_EOS:EOS_SWATH:"C:\AMSR_E_L2_Ocean_B01_200206182340_A.hdf":Swath1:Low_res_sst',
-            r"C:\AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
-        ),
-        (
-            "HDF4_EOS:EOS_SWATH:AMSR_E_L2_Ocean_B01_200206182340_A.hdf:Swath1:Low_res_sst",
-            "AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
-        ),
-        (
-            r"HDF4_EOS:EOS_SWATH:C:\AMSR_E_L2_Ocean_B01_200206182340_A.hdf:Swath1:Low_res_sst",
-            r"C:\AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
-        ),
-        (
-            r"HDF4_EOS:EOS_SWATH:C:/AMSR_E_L2_Ocean_B01_200206182340_A.hdf:Swath1:Low_res_sst",
-            r"C:/AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
-        ),
-        (
             r'HDF4_EOS:EOS_SWATH:"/vsicurl/http://www.my.com/OMI-Aura_L2-OMTO3_2005m0113t0224-o02648_v002-2005m0625t035355.he5":Swath1:Low_res_sst',
             r"/vsicurl/http://www.my.com/OMI-Aura_L2-OMTO3_2005m0113t0224-o02648_v002-2005m0625t035355.he5",
         ),
-        (
-            r"HDF4_EOS:EOS_SWATH:a:Swath1:Low_res_sst",
-            r"a",
-        ),
-        ("", ""),
     ),
 )
 def test_gdal_subdataset_get_filename(filename, path_component):

--- a/autotest/gdrivers/hdf5.py
+++ b/autotest/gdrivers/hdf5.py
@@ -1404,6 +1404,22 @@ def test_gdal_subdataset_get_filename(filename, path_component):
 
 @pytest.mark.parametrize(
     "filename",
+    ("/vsimem/DATA:TEST.h5", "/vsimem/DATA:TEST:DOUBLE.h5", '/vsimem/DATA:"TEST".h5'),
+)
+def test_gdal_subdataset_get_filename_colon(filename):
+    """Test regression #14978"""
+
+    gdal.FileFromMemBuffer(
+        filename, open("data/hdf5/hdfeos_sample_swath.h5", "rb").read()
+    )
+    info = gdal.GetSubdatasetInfo(f'HDF5:"{filename}":/navigation_data/latitude')
+    assert info.GetPathComponent() == filename
+    assert info.GetSubdatasetComponent() == "/navigation_data/latitude"
+    gdal.Unlink(filename)
+
+
+@pytest.mark.parametrize(
+    "filename",
     (
         'HDF5:"OMI-Aura_L2-OMTO3_2005m0113t0224-o02648_v002-2005m0625t035355.he5"://HDFEOS/SWATHS/OMI_Column_Amount_O3/Data_Fields/UVAerosolIndex',
         r'HDF5:"C:\OMI-Aura_L2-OMTO3_2005m0113t0224-o02648_v002-2005m0625t035355.he5"://HDFEOS/SWATHS/OMI_Column_Amount_O3/Data_Fields/UVAerosolIndex',

--- a/autotest/gdrivers/netcdf.py
+++ b/autotest/gdrivers/netcdf.py
@@ -6526,6 +6526,23 @@ def test_gdal_subdataset_get_filename(filename, path_component, subdataset_compo
 
 @pytest.mark.parametrize(
     "filename",
+    ("/vsimem/DATA:TEST.nc", "/vsimem/DATA:TEST:DOUBLE.nc", '/vsimem/DATA:"TEST".nc'),
+)
+def test_gdal_subdataset_get_filename_colon(filename):
+    """Test regression #14978"""
+
+    gdal.FileFromMemBuffer(
+        filename,
+        open("data/netcdf/fake_SNPP_VIIRS.20230406T024200.L2.OC.NRT.nc", "rb").read(),
+    )
+    info = gdal.GetSubdatasetInfo(f'NETCDF:"{filename}":/navigation_data/latitude')
+    assert info.GetPathComponent() == filename
+    assert info.GetSubdatasetComponent() == "/navigation_data/latitude"
+    gdal.Unlink(filename)
+
+
+@pytest.mark.parametrize(
+    "filename",
     (
         "NETCDF:",
         'NETCDF:"data/netcdf/SNPP_VIIRS.20230406T024200.L2.OC.NRT.nc":/navigation_data/longitude',

--- a/frmts/hdf4/hdf4drivercore.cpp
+++ b/frmts/hdf4/hdf4drivercore.cpp
@@ -64,6 +64,16 @@ void HDF4DriverSubdatasetInfo::parseFileName()
     CPLStringList aosParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
     const int iPartsCount{CSLCount(aosParts)};
 
+    auto unescapeDoubleQuotes = [](std::string &str)
+    {
+        size_t pos = 0;
+        while ((pos = str.find("\\\"", pos)) != std::string::npos)
+        {
+            str.replace(pos, 2, "\"");
+            pos += 1;
+        }
+    };
+
     if (iPartsCount >= 3)
     {
 
@@ -76,6 +86,38 @@ void HDF4DriverSubdatasetInfo::parseFileName()
 
         if (iPartsCount >= 4)
         {
+
+            std::string part2{aosParts[2]};
+            const bool pathIsDoubleQuoted{!part2.empty() && part2[0] == '"'};
+
+            if (pathIsDoubleQuoted)
+            {
+                // The path component is everything up to the next double quote.
+                if (part2.back() != '"')
+                {
+                    // If the path component is double quoted and there is no closing quote, then the path component
+                    // is everything up to the next part that ends with an unescaped double quote.
+                    // This is to handle cases where the path component contains colons.
+                    for (int i = 3; i < iPartsCount; ++i)
+                    {
+                        const int partLen = strlen(aosParts[i]);
+                        if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
+                            !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
+                        {
+                            part2.append(":");
+                            part2.append(std::string(aosParts[i]));
+                            subdatasetIndex = i + 1;
+                            break;
+                        }
+                        part2.append(":");
+                        part2.append(std::string(aosParts[i]));
+                    }
+                }
+                unescapeDoubleQuotes(part2);
+            }
+
+            m_pathComponent = part2;
+
             const bool hasDriveLetter{
                 (strlen(aosParts[3]) > 1 &&
                  (aosParts[3][0] == '\\' || aosParts[3][0] == '/')) &&
@@ -83,12 +125,11 @@ void HDF4DriverSubdatasetInfo::parseFileName()
                   std::isalpha(static_cast<unsigned char>(aosParts[2][1]))) ||
                  (strlen(aosParts[2]) == 1 &&
                   std::isalpha(static_cast<unsigned char>(aosParts[2][0]))))};
-            m_pathComponent = aosParts[2];
 
             const bool hasProtocol{m_pathComponent.find("/vsicurl/") !=
                                    std::string::npos};
 
-            if (hasDriveLetter || hasProtocol)
+            if (!pathIsDoubleQuoted && (hasDriveLetter || hasProtocol))
             {
                 m_pathComponent.append(":");
                 m_pathComponent.append(aosParts[3]);

--- a/frmts/hdf4/hdf4drivercore.cpp
+++ b/frmts/hdf4/hdf4drivercore.cpp
@@ -100,7 +100,7 @@ void HDF4DriverSubdatasetInfo::parseFileName()
                     // This is to handle cases where the path component contains colons.
                     for (int i = 3; i < iPartsCount; ++i)
                     {
-                        const int partLen = strlen(aosParts[i]);
+                        const size_t partLen = strlen(aosParts[i]);
                         if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
                             !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
                         {

--- a/frmts/hdf4/hdf4drivercore.cpp
+++ b/frmts/hdf4/hdf4drivercore.cpp
@@ -61,18 +61,9 @@ void HDF4DriverSubdatasetInfo::parseFileName()
         return;
     }
 
-    CPLStringList aosParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
+    CPLStringList aosParts{CSLTokenizeString2(
+        m_fileName.c_str(), ":", CSLT_HONOURSTRINGS | CSLT_PRESERVEQUOTES)};
     const int iPartsCount{CSLCount(aosParts)};
-
-    auto unescapeDoubleQuotes = [](std::string &str)
-    {
-        size_t pos = 0;
-        while ((pos = str.find("\\\"", pos)) != std::string::npos)
-        {
-            str.replace(pos, 2, "\"");
-            pos += 1;
-        }
-    };
 
     if (iPartsCount >= 3)
     {
@@ -86,38 +77,6 @@ void HDF4DriverSubdatasetInfo::parseFileName()
 
         if (iPartsCount >= 4)
         {
-
-            std::string part2{aosParts[2]};
-            const bool pathIsDoubleQuoted{!part2.empty() && part2[0] == '"'};
-
-            if (pathIsDoubleQuoted)
-            {
-                // The path component is everything up to the next double quote.
-                if (part2.back() != '"')
-                {
-                    // If the path component is double quoted and there is no closing quote, then the path component
-                    // is everything up to the next part that ends with an unescaped double quote.
-                    // This is to handle cases where the path component contains colons.
-                    for (int i = 3; i < iPartsCount; ++i)
-                    {
-                        const size_t partLen = strlen(aosParts[i]);
-                        if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
-                            !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
-                        {
-                            part2.append(":");
-                            part2.append(std::string(aosParts[i]));
-                            subdatasetIndex = i + 1;
-                            break;
-                        }
-                        part2.append(":");
-                        part2.append(std::string(aosParts[i]));
-                    }
-                }
-                unescapeDoubleQuotes(part2);
-            }
-
-            m_pathComponent = part2;
-
             const bool hasDriveLetter{
                 (strlen(aosParts[3]) > 1 &&
                  (aosParts[3][0] == '\\' || aosParts[3][0] == '/')) &&
@@ -125,11 +84,16 @@ void HDF4DriverSubdatasetInfo::parseFileName()
                   std::isalpha(static_cast<unsigned char>(aosParts[2][1]))) ||
                  (strlen(aosParts[2]) == 1 &&
                   std::isalpha(static_cast<unsigned char>(aosParts[2][0]))))};
+            m_pathComponent = aosParts[2];
 
             const bool hasProtocol{m_pathComponent.find("/vsicurl/") !=
                                    std::string::npos};
 
-            if (!pathIsDoubleQuoted && (hasDriveLetter || hasProtocol))
+            // If the patch component was double quoted, the drive or
+            // protocol will be part of the path component and we should not append it again
+
+            if ((m_pathComponent.size() > 0 && m_pathComponent[0] != '"') &&
+                (hasDriveLetter || hasProtocol))
             {
                 m_pathComponent.append(":");
                 m_pathComponent.append(aosParts[3]);

--- a/frmts/hdf5/hdf5drivercore.cpp
+++ b/frmts/hdf5/hdf5drivercore.cpp
@@ -209,7 +209,7 @@ void HDF5DriverSubdatasetInfo::parseFileName()
                 // This is to handle cases where the path component contains colons.
                 for (int i = 2; i < iPartsCount; ++i)
                 {
-                    const int partLen = strlen(aosParts[i]);
+                    const size_t partLen = strlen(aosParts[i]);
                     if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
                         !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
                     {

--- a/frmts/hdf5/hdf5drivercore.cpp
+++ b/frmts/hdf5/hdf5drivercore.cpp
@@ -177,54 +177,22 @@ void HDF5DriverSubdatasetInfo::parseFileName()
         return;
     }
 
-    CPLStringList aosParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
+    CPLStringList aosParts{CSLTokenizeString2(
+        m_fileName.c_str(), ":", CSLT_HONOURSTRINGS | CSLT_PRESERVEQUOTES)};
     const int iPartsCount{CSLCount(aosParts)};
-
-    auto unescapeDoubleQuotes = [](std::string &str)
-    {
-        size_t pos = 0;
-        while ((pos = str.find("\\\"", pos)) != std::string::npos)
-        {
-            str.replace(pos, 2, "\"");
-            pos += 1;
-        }
-    };
 
     if (iPartsCount >= 3)
     {
 
-        std::string part1{aosParts[1]};
-        const bool pathIsDoubleQuoted{!part1.empty() && part1[0] == '"'};
-
         m_driverPrefixComponent = aosParts[0];
-        int subdatasetIndex{2};
 
-        if (pathIsDoubleQuoted)
+        std::string part1{aosParts[1]};
+        if (!part1.empty() && part1[0] == '"')
         {
-            // The path component is everything up to the next double quote.
-            if (part1.back() != '"')
-            {
-                // If the path component is double quoted and there is no closing quote, then the path component
-                // is everything up to the next part that ends with an unescaped double quote.
-                // This is to handle cases where the path component contains colons.
-                for (int i = 2; i < iPartsCount; ++i)
-                {
-                    const size_t partLen = strlen(aosParts[i]);
-                    if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
-                        !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
-                    {
-                        part1.append(":");
-                        part1.append(std::string(aosParts[i]));
-                        subdatasetIndex = i + 1;
-                        break;
-                    }
-                    part1.append(":");
-                    part1.append(std::string(aosParts[i]));
-                }
-            }
-            unescapeDoubleQuotes(part1);
+            part1 = part1.substr(1);
         }
 
+        int subdatasetIndex{2};
         const bool hasDriveLetter{
             part1.length() == 1 &&
             std::isalpha(static_cast<unsigned char>(part1.at(0))) &&
@@ -237,7 +205,7 @@ void HDF5DriverSubdatasetInfo::parseFileName()
                                part1 == "/vsicurl_streaming/http" ||
                                part1 == "/vsicurl_streaming/https"};
 
-        m_pathComponent = part1;
+        m_pathComponent = aosParts[1];
 
         if (hasDriveLetter || hasProtocol)
         {

--- a/frmts/hdf5/hdf5drivercore.cpp
+++ b/frmts/hdf5/hdf5drivercore.cpp
@@ -180,18 +180,51 @@ void HDF5DriverSubdatasetInfo::parseFileName()
     CPLStringList aosParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
     const int iPartsCount{CSLCount(aosParts)};
 
+    auto unescapeDoubleQuotes = [](std::string &str)
+    {
+        size_t pos = 0;
+        while ((pos = str.find("\\\"", pos)) != std::string::npos)
+        {
+            str.replace(pos, 2, "\"");
+            pos += 1;
+        }
+    };
+
     if (iPartsCount >= 3)
     {
 
-        m_driverPrefixComponent = aosParts[0];
-
         std::string part1{aosParts[1]};
-        if (!part1.empty() && part1[0] == '"')
+        const bool pathIsDoubleQuoted{!part1.empty() && part1[0] == '"'};
+
+        m_driverPrefixComponent = aosParts[0];
+        int subdatasetIndex{2};
+
+        if (pathIsDoubleQuoted)
         {
-            part1 = part1.substr(1);
+            // The path component is everything up to the next double quote.
+            if (part1.back() != '"')
+            {
+                // If the path component is double quoted and there is no closing quote, then the path component
+                // is everything up to the next part that ends with an unescaped double quote.
+                // This is to handle cases where the path component contains colons.
+                for (int i = 2; i < iPartsCount; ++i)
+                {
+                    const int partLen = strlen(aosParts[i]);
+                    if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
+                        !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
+                    {
+                        part1.append(":");
+                        part1.append(std::string(aosParts[i]));
+                        subdatasetIndex = i + 1;
+                        break;
+                    }
+                    part1.append(":");
+                    part1.append(std::string(aosParts[i]));
+                }
+            }
+            unescapeDoubleQuotes(part1);
         }
 
-        int subdatasetIndex{2};
         const bool hasDriveLetter{
             part1.length() == 1 &&
             std::isalpha(static_cast<unsigned char>(part1.at(0))) &&
@@ -204,7 +237,7 @@ void HDF5DriverSubdatasetInfo::parseFileName()
                                part1 == "/vsicurl_streaming/http" ||
                                part1 == "/vsicurl_streaming/https"};
 
-        m_pathComponent = aosParts[1];
+        m_pathComponent = part1;
 
         if (hasDriveLetter || hasProtocol)
         {

--- a/frmts/netcdf/netcdfdrivercore.cpp
+++ b/frmts/netcdf/netcdfdrivercore.cpp
@@ -276,7 +276,7 @@ void NCDFDriverSubdatasetInfo::parseFileName()
             // This is to handle cases where the path component contains colons.
             for (int i = 2; i < iPartsCount; ++i)
             {
-                const int partLen = strlen(aosParts[i]);
+                const size_t partLen = strlen(aosParts[i]);
                 if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
                     !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
                 {

--- a/frmts/netcdf/netcdfdrivercore.cpp
+++ b/frmts/netcdf/netcdfdrivercore.cpp
@@ -236,7 +236,8 @@ void NCDFDriverSubdatasetInfo::parseFileName()
         return;
     }
 
-    CPLStringList aosParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
+    CPLStringList aosParts{CSLTokenizeString2(
+        m_fileName.c_str(), ":", CSLT_HONOURSTRINGS | CSLT_PRESERVEQUOTES)};
     const int iPartsCount{CSLCount(aosParts)};
 
     m_driverPrefixComponent = aosParts[0];
@@ -251,45 +252,12 @@ void NCDFDriverSubdatasetInfo::parseFileName()
         return;
     }
 
-    auto unescapeDoubleQuotes = [](std::string &str)
-    {
-        size_t pos = 0;
-        while ((pos = str.find("\\\"", pos)) != std::string::npos)
-        {
-            str.replace(pos, 2, "\"");
-            pos += 1;
-        }
-    };
-
     int subdatasetIndex{2};
 
     std::string part1{aosParts[1]};
-    const bool pathIsDoubleQuoted{!part1.empty() && part1[0] == '"'};
-
-    if (pathIsDoubleQuoted)
+    if (!part1.empty() && part1[0] == '"')
     {
-        // The path component is everything up to the next double quote.
-        if (part1.back() != '"')
-        {
-            // If the path component is double quoted and there is no closing quote, then the path component
-            // is everything up to the next part that ends with an unescaped double quote.
-            // This is to handle cases where the path component contains colons.
-            for (int i = 2; i < iPartsCount; ++i)
-            {
-                const size_t partLen = strlen(aosParts[i]);
-                if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
-                    !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
-                {
-                    part1.append(":");
-                    part1.append(std::string(aosParts[i]));
-                    subdatasetIndex = i + 1;
-                    break;
-                }
-                part1.append(":");
-                part1.append(std::string(aosParts[i]));
-            }
-        }
-        unescapeDoubleQuotes(part1);
+        part1 = part1.substr(1);
     }
 
     const bool hasDriveLetter{
@@ -304,8 +272,7 @@ void NCDFDriverSubdatasetInfo::parseFileName()
                            part1 == "/vsicurl_streaming/https" ||
                            part1 == "http" || part1 == "https"};
 
-    m_pathComponent = part1;
-
+    m_pathComponent = aosParts[1];
     if (hasDriveLetter || hasProtocol)
     {
         m_pathComponent.append(":");
@@ -348,8 +315,8 @@ void NCDFDriverSubdatasetInfo::parseFileName()
     {
         m_subdatasetComponent = m_subdatasetComponent.substr(1);
     }
-
-    if (!m_subdatasetComponent.empty() && m_subdatasetComponent.back() == '"')
+    if (!m_subdatasetComponent.empty() &&
+        m_subdatasetComponent.rfind('"') == m_subdatasetComponent.length() - 1)
     {
         m_subdatasetComponent.pop_back();
     }

--- a/frmts/netcdf/netcdfdrivercore.cpp
+++ b/frmts/netcdf/netcdfdrivercore.cpp
@@ -251,12 +251,45 @@ void NCDFDriverSubdatasetInfo::parseFileName()
         return;
     }
 
+    auto unescapeDoubleQuotes = [](std::string &str)
+    {
+        size_t pos = 0;
+        while ((pos = str.find("\\\"", pos)) != std::string::npos)
+        {
+            str.replace(pos, 2, "\"");
+            pos += 1;
+        }
+    };
+
     int subdatasetIndex{2};
 
     std::string part1{aosParts[1]};
-    if (!part1.empty() && part1[0] == '"')
+    const bool pathIsDoubleQuoted{!part1.empty() && part1[0] == '"'};
+
+    if (pathIsDoubleQuoted)
     {
-        part1 = part1.substr(1);
+        // The path component is everything up to the next double quote.
+        if (part1.back() != '"')
+        {
+            // If the path component is double quoted and there is no closing quote, then the path component
+            // is everything up to the next part that ends with an unescaped double quote.
+            // This is to handle cases where the path component contains colons.
+            for (int i = 2; i < iPartsCount; ++i)
+            {
+                const int partLen = strlen(aosParts[i]);
+                if (partLen > 0 && aosParts[i][partLen - 1] == '"' &&
+                    !(partLen > 1 && aosParts[i][partLen - 2] == '\\'))
+                {
+                    part1.append(":");
+                    part1.append(std::string(aosParts[i]));
+                    subdatasetIndex = i + 1;
+                    break;
+                }
+                part1.append(":");
+                part1.append(std::string(aosParts[i]));
+            }
+        }
+        unescapeDoubleQuotes(part1);
     }
 
     const bool hasDriveLetter{
@@ -271,7 +304,8 @@ void NCDFDriverSubdatasetInfo::parseFileName()
                            part1 == "/vsicurl_streaming/https" ||
                            part1 == "http" || part1 == "https"};
 
-    m_pathComponent = aosParts[1];
+    m_pathComponent = part1;
+
     if (hasDriveLetter || hasProtocol)
     {
         m_pathComponent.append(":");
@@ -314,8 +348,8 @@ void NCDFDriverSubdatasetInfo::parseFileName()
     {
         m_subdatasetComponent = m_subdatasetComponent.substr(1);
     }
-    if (!m_subdatasetComponent.empty() &&
-        m_subdatasetComponent.rfind('"') == m_subdatasetComponent.length() - 1)
+
+    if (!m_subdatasetComponent.empty() && m_subdatasetComponent.back() == '"')
     {
         m_subdatasetComponent.pop_back();
     }

